controllers/client/grafana_client.go

```diff
@@ -17,12 +21,82 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+const (
+	serviceAccountTokenPath = "/var/run/secrets/kubernetes.io/serviceaccount/token" // nolint:gosec
+)
+
 type grafanaAdminCredentials struct {
 	adminUser     string
 	adminPassword string
 	apikey        string
 }
 
+type JWTCache struct {
+	Token      string
+	Expiration time.Time
+}
+
+var jwtCache *JWTCache
+
+// Revoke tokens early expecting them to be rotated hourly, see 'ExpirationSeconds' in KEP1205
+// Should mitigate mid-reconcile expiration
+const tokenExpirationCompensation = -30 * time.Second
+
+// getBearerToken will read JWT token from given file and cache it until it expires.
+// accepts filepath arg for testing
+func getBearerToken(bearerTokenPath string) (string, error) {
+	// Return cached token if not expired
+	if jwtCache != nil && jwtCache.Expiration.After(time.Now()) {
+		return jwtCache.Token, nil
+	}
+
+	b, err := os.ReadFile(bearerTokenPath)
+	if err != nil {
+		return "", fmt.Errorf("reading token file at %s, %w", bearerTokenPath, err)
+	}
+
+	token := string(b)
+
+	encClaims := strings.Split(token, ".")
+	if len(encClaims) != 3 {
+		return "", fmt.Errorf("ServiceAccount JWT token expected to have 3 parts, not %d", len(encClaims))
+	}
+
+	rawClaims, err := base64.RawStdEncoding.DecodeString(encClaims[1])
+	if err != nil {
+		return "", fmt.Errorf("base64 decoding ServiceAccount JWT token %w", err)
+	}
+
+	var claims map[string]any
+
+	err = json.Unmarshal(rawClaims, &claims)
+	if err != nil {
+		return "", fmt.Errorf("deserializing ServiceAccount JWT claims %w", err)
+	}
+
+	rawExp, ok := claims["exp"]
+	if !ok {
+		return "", fmt.Errorf("no expiry found in ServiceAccount JWT claims")
+	}
+
+	exp, ok := rawExp.(float64)
+	if !ok {
+		return "", fmt.Errorf("token exp claim (expiry) cannot be cast to a float64")
+	}
+
+	tokenExpiration := time.Unix(int64(exp), 0)
+	if tokenExpiration.Add(tokenExpirationCompensation).Before(time.Now()) {
+		return "", fmt.Errorf("token expired at %s, expected %s to be renewed. Tokens are considered expired 30 seconds early", tokenExpiration.String(), bearerTokenPath)
```

Comment:
does it make sense to do an exponential backoff here, retrying until the token is no longer expired?

Reply:
Yeah no it makes sense - I didn't consider the time scale we're working on here so this should be good!